numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -4610,20 +4509,75 @@ set_typeinfo(PyObject *dict)
      *          CFLOAT, CDOUBLE, CLONGDOUBLE,
      *          OBJECT, STRING, UNICODE, VOID,
      *          DATETIME,TIMEDELTA#
+     * #NAME = Bool,
+     *         Byte, UByte, Short, UShort, Int, UInt,
+     *         Long, ULong, LongLong, ULongLong,
+     *         Half, Float, Double, LongDouble,
+     *         CFloat, CDouble, CLongDouble,
+     *         Object, String, Unicode, Void,
+     *         Datetime, Timedelta#
+     * #kind = GENBOOL,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         FLOATING, FLOATING, FLOATING, FLOATING,
+     *         COMPLEX, COMPLEX, COMPLEX,
+     *         OBJECT, STRING, UNICODE, VOID,
+     *         DATETIME, TIMEDELTA#
+     * #endian = |*3, =*15, |, |, =, |, =*2#
+     * #flags= 0*18, NPY_OBJECT_DTYPE_FLAGS,0, NPY_NEEDS_INIT, 0*3#
+     * #fromtype = npy_bool,
+     *             npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+     *             npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+     *             npy_half, npy_float, npy_double, npy_longdouble,
+     *             npy_cfloat, npy_cdouble, npy_clongdouble,
+     *             PyObject *, char, npy_ucs4, char,
+     *             npy_datetime, npy_timedelta#
+     * #DO_ELSIZE = 1*19, 0*3, 1*2#
      */
 
-    /**begin repeat1
-     *
-     * #name2 = HALF, DATETIME, TIMEDELTA#
-     */
+#if USE_DTYPE_AS_PYOBJECT
+    dtype = PyObject_New(PyArray_Descr, &PyArrayDescr_Type);
+#else
+    dtype = PyObject_GC_New(PyArray_Descr, &PyArrayDescr_Type);
+#endif
+    /* Don't copy PyObject_HEAD part */
+    memcpy((char *)dtype + sizeof(PyObject),
+           (char *)&PyArrayDescr_Type + sizeof(PyObject),
+           sizeof(PyArray_Descr) - sizeof(PyObject));
```

Comment:
I don't really follow why this is needed. What is being copied here?

Reply:
This is duplicated from `PyArray_DescrNew`, and copies the type object as well as the dtype-specific values `kind`, `itemsize`, ...

Reply:
Aren't you setting all the dtype-specific values below?

Reply:
I don't think that's true at all. The metatype object and the type objects should have nothing in common. I think this is what is breaking your classmethods.

Reply:
As an example, `dtype->type->tp_new` should be set to a function raising `NotImplementedError`

Reply:
I'm labelling this as resolved since it already led to confusion!